lisa/transformers/upgrade_packages.py

```diff
@@ -0,0 +1,170 @@
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return UpgradeInstallerTransformerSchema
+
+    @property
+    def _output_names(self) -> List[str]:
+        return []
+
+    def _internal_run(self) -> Dict[str, Any]:
+        runbook: UpgradeInstallerTransformerSchema = self.runbook
+        assert runbook.connection, "connection must be defined."
+        assert runbook.installer, "installer must be defined."
+
+        node = quick_connect(runbook.connection, "installer_node")
+
+        uname = node.tools[Uname]
+        self._log.info(
+            f"kernel version before install: {uname.get_linux_information()}"
+        )
+        factory = subclasses.Factory[BaseInstaller](BaseInstaller)
+        installer = factory.create_by_runbook(
+            runbook=runbook.installer, node=node, parent_log=self._log
+        )
+
+        installer.validate()
+        package_list = installer.install()
+        self._log.info(f"Packages upgraded: {package_list}")
+
+        self._log.info("rebooting")
+        node.reboot()
+        self._log.info(
+            f"kernel version after install: "
+            f"{uname.get_linux_information(force_run=True)}"
+        )
+
+        return {}
+
+
+class UnattendedUpgradeInstaller(BaseInstaller):
+    def __init__(
+        self,
+        runbook: Any,
+        node: Node,
+        parent_log: Logger,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(runbook, node, parent_log, *args, **kwargs)
+        self.repo_url = "http://archive.ubuntu.com/ubuntu/"
+
+    @classmethod
+    def type_name(cls) -> str:
+        return "unattended_upgrade"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return BaseInstallerSchema
+
+    def validate(self) -> None:
+        assert isinstance(self._node.os, Debian), (
+            f"The '{self.type_name()}' installer only supports Debian family. "
+            f"The current os is {self._node.os.name}"
+        )
+
+    def install(self) -> str:
+        node: Node = self._node
+        assert isinstance(node.os, Debian)
+
+        cmd_result = node.execute(
+            "which unattended-upgrade",
+            sudo=True,
+            shell=True,
+        )
+        if 0 != cmd_result.exit_code:
+            node.os.install_packages("unattended-upgrades")
+        if type(node.os) == Debian:
+            if node.os.information.version >= "10.0.0":
+                node.execute(
+                    "mkdir -p /var/cache/apt/archives/partial",
+                    sudo=True,
+                    shell=True,
+                    expected_exit_code=0,
+                    expected_exit_code_failure_message=(
+                        "fail to make folder /var/cache/apt/archives/partial"
+                    ),
+                )
+            else:
+                node.os.install_packages(["debian-keyring", "debian-archive-keyring"])
+        node.execute(
```

Comment:
I think most methods are implemented in os and tools. The transformers are higher level classes, so they can reference tools and os as wanted.